File: chroma-league/src/main/java/com/bonepl/chromaleague/hud/parts/health/HealthBar.java
package com.bonepl.chromaleague.hud.parts.health;

import com.bonepl.chromaleague.hud.colors.CLColor;
import com.bonepl.chromaleague.hud.parts.ProgressBar;
import com.bonepl.chromaleague.state.GameStateHelper;
import com.bonepl.chromaleague.state.RunningState;
import com.bonepl.razersdk.animation.Animation;
import com.bonepl.razersdk.animation.Frame;
import com.bonepl.razersdk.animation.IFrame;
import com.bonepl.razersdk.sdk.RzKey;
import com.bonepl.razersdk.sdk.RzKeySelector;

import java.util.Collections;
import java.util.List;

import static com.bonepl.razersdk.sdk.RzKey.RZKEY_ESC;
import static com.bonepl.razersdk.sdk.RzKey.RZKEY_F12;

public class HealthBar extends Animation {
    private static final List<RzKey> HP_BAR_KEYS = new RzKeySelector()
            .withRowOf(RZKEY_ESC)
            .withColumnBetween(RZKEY_ESC, RZKEY_F12)
            .sortedByColumn()
            .asList();
    private double previousHp;

    @Override
    public Frame getFrame() {
        final double currentHp = RunningState.getGameState().getActivePlayer().championStats().currentHealth();
        addToBack(getHpBar());
        if (currentHp < previousHp) {
            addToFront(new LostHealthAnimation(previousHp, currentHp));
        }
        if (currentHp > previousHp) {
            addToFront(new GainedHealthAnimation(previousHp, currentHp));
        }
        previousHp = currentHp;
        return super.getFrame();
    }

    private static IFrame getHpBar() {
        return new ProgressBar(HP_BAR_KEYS, GameStateHelper.getHpPercentage(), CLColor.HEALTH);
    }

    public static List<RzKey> getHealthBarKeys() {
        return Collections.unmodifiableList(HP_BAR_KEYS);
    }

    public static List<RzKey> getHpBarPart(double previousHp, double currentHp) {
        final double maxHealth = RunningState.getGameState().getActivePlayer().championStats().maxHealth();
        final int from = ProgressBar.indexToFill(getHealthBarKeys(), Double.valueOf(previousHp * 100 / maxHealth).intValue());
        final int to = ProgressBar.indexToFill(getHealthBarKeys(), Double.valueOf(currentHp * 100 / maxHealth).intValue());
        return getHealthBarKeys().subList(from, to);
    }
}

File: src/hud/parts/health/HealthBar.ts
import {Animation} from '../../../razer-sdk/animation/Animation.js';
import {Frame} from '../../../razer-sdk/animation/Frame.js';
import {IFrame} from '../../../razer-sdk/animation/IFrame.js';
import {RzKey} from '../../../razer-sdk/sdk/RzKey.js';
import {RzKeySelector} from '../../../razer-sdk/sdk/RzKeySelector.js';
import * as GameStateHelper from '../../../state/GameStateHelper.js';
import {RunningState} from '../../../state/RunningState.js';
import {CLColor} from '../../colors/CLColor.js';
import {indexToFill, ProgressBar} from '../ProgressBar.js';
import {GainedHealthAnimation} from './GainedHealthAnimation.js';
import {LostHealthAnimation} from './LostHealthAnimation.js';

const HP_BAR_KEYS: RzKey[] = new RzKeySelector()
    .withRowOf(RzKey.RZKEY_ESC)
    .withColumnBetween(RzKey.RZKEY_ESC, RzKey.RZKEY_F12)
    .sortedByColumn()
    .asList();

export class HealthBar extends Animation {
    private previousHp = 0;

    override pollFrame(): Frame {
        const currentHp = RunningState.getGameState().activePlayer!.championStats.currentHealth;
        this.addToBack(getHpBar());
        if (currentHp < this.previousHp) {
            this.addToFront(new LostHealthAnimation(this.previousHp, currentHp));
        }
        if (currentHp > this.previousHp) {
            this.addToFront(new GainedHealthAnimation(this.previousHp, currentHp));
        }
        this.previousHp = currentHp;
        return super.pollFrame();
    }
}

function getHpBar(): IFrame {
    return new ProgressBar(HP_BAR_KEYS, GameStateHelper.getHpPercentage(), CLColor.HEALTH);
}

export function getHealthBarKeys(): RzKey[] {
    return [...HP_BAR_KEYS];
}

export function getHpBarPart(previousHp: number, currentHp: number): RzKey[] {
    const maxHealth = RunningState.getGameState().activePlayer!.championStats.maxHealth;
    const from = indexToFill(getHealthBarKeys().length, Math.trunc(previousHp * 100 / maxHealth));
    const to = indexToFill(getHealthBarKeys().length, Math.trunc(currentHp * 100 / maxHealth));
    return getHealthBarKeys().slice(from, to);
}
